prov/gni/src/gnix_mr.c

```diff
@@ -771,127 +973,292 @@ int _gnix_mr_cache_flush(gnix_mr_cache_t *cache)
 	return FI_SUCCESS;
 }
 
-static int __mr_cache_lookup_inuse_fastpath(
-		gnix_mr_cache_t *cache,
-		gnix_mr_cache_key_t *key,
-		gnix_mr_cache_entry_t **entry)
+static int __mr_cache_search_inuse(
+		gnix_mr_cache_t          *cache,
+		struct gnix_fid_domain   *domain,
+		uint64_t                 address,
+		uint64_t                 length,
+		gni_cq_handle_t          dst_cq_hndl,
+		uint32_t                 flags,
+		uint32_t                 vmdh_index,
+		gnix_mr_cache_entry_t    **entry,
+		gnix_mr_cache_key_t      *key)
 {
+	int ret = FI_SUCCESS, cmp;
 	RbtIterator iter;
-	int ret = -FI_ENOENT;
-	gnix_mr_cache_key_t *e_key;
+	RbtStatus rc;
+	gnix_mr_cache_key_t *found_key, new_key;
+	gnix_mr_cache_entry_t *found_entry, *next_entry;
+	uint64_t new_end, found_end;
+	DLIST_HEAD(tmp);
+
+	/* first we need to find an entry that overlaps with this one.
+	 * care should be taken to find the left most entry that overlaps
+	 * with this entry since the entry we are searching for might overlap
+	 * many entries and the tree may be left or right balanced
+	 * at the head
+	 */
+	iter = rbtFindLeftmost(cache->inuse.rb_tree, (void *) key,
+			__find_overlapping_addr);
+	if (!iter) {
+		GNIX_INFO(FI_LOG_MR,
+				"could not find key in inuse, key=%llu:%llu\n",
+				key->address, key->length);
+		return -FI_ENOENT;
+	}
 
-	*entry = NULL;
+	rbtKeyValue(cache->inuse.rb_tree, iter, (void **) &found_key,
+			(void **) &found_entry);
+
+	GNIX_INFO(FI_LOG_MR,
+			"found a key that matches the search criteria, "
+			"found=%llu:%llu key=%llu:%llu\n",
+			found_key->address, found_key->length,
+			key->address, key->length);
+	/* if the entry that we've found completely subsumes
+	 * the requested entry, just return a reference to
+	 * that existing registration
+	 */
+	if (__can_subsume(found_key, key)) {
+		GNIX_INFO(FI_LOG_MR,
+				"found an entry that subsumes the request, "
+				"existing=%llu:%llu key=%llu:%llu\n",
+				found_key->address, found_key->length,
+				key->address, key->length);
+		*entry = found_entry;
+		__mr_cache_entry_get(cache, found_entry);
+
+		cache->hits++;
+		return FI_SUCCESS;
+	}
 
-	/* Is the key in the inuse tree? */
-	iter = rbtFind(cache->inuse, key);
-	if (iter) {
-		/* let's find a matching element */
-		rbtKeyValue(cache->inuse, iter, (void **) &e_key,
-				(void **) entry);
+	/* otherwise, iterate from the existing entry until we can no longer
+	 * find an entry that overlaps with the new registration and remove
+	 * and retire each of those entries.
+	 */
+	new_key.address = MIN(found_key->address, key->address);
+	new_end = key->address + key->length;
```

Comment:
If the found key has a lower base address and extends past the requested
region, then it would fully subsume the request and would not reach this
point. 
-- Jim




On 1/6/16, 1:09 PM, "Zach Tiffany" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_mr.c
><https://github.com/ofi-cray/libfabric-cray/pull/582#discussion_r48995660>
>:
>>  
>> -	/* Is the key in the inuse tree? */
>> -	iter = rbtFind(cache->inuse, key);
>> -	if (iter) {
>> -		/* let's find a matching element */
>> -		rbtKeyValue(cache->inuse, iter, (void **) &e_key,
>> -				(void **) entry);
>> +	/* otherwise, iterate from the existing entry until we can no longer
>> +	 * find an entry that overlaps with the new registration and remove
>> +	 * and retire each of those entries.
>> +	 */
>> +	new_key.address = MIN(found_key->address, key->address);
>> +	new_end = key->address + key->length;
>Couldn't the "found_key" registration extend past the requested region?
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/582/files#r48995660>.
>
>